075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,435 @@
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration.
+
+In the following example, a user can decide to use the same rebalancing configuration when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the same rebalancing configuration when adding or removing brokers
+      - mode: add-brokers
+        config: my-add-remove-brokers-rebalancing-config
+      - mode: remove-brokers
+        config: my-add-remove-brokers-rebalancing-config
+```
+
+Furthermore, it is possible to use the default Cruise Control rebalancing configuration by omitting the corresponding field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the default Cruise Control rebalancing configuration when adding or removing brokers
+      - mode: add-brokers
+      - mode: remove-brokers
+```
+
+The user can even decide to use a different configuration when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using specific configuration when adding brokers
+      - mode: add-brokers
+        config: my-add-brokers-rebalancing-config
+      # using different configuration when removing brokers  
+      - mode: remove-brokers
+        config: my-remove-brokers-rebalancing-config
+```
+
+The user could even decide to have auto-rebalancing only when adding or removing brokers but not for both.
+This way provides the greatest flexibility to describe in which case to run the auto-rebalancing and with which configuration.
+
+The auto-rebalance configuration for the `spec.cruiseControl.autoRebalance.config` property in the `Kafka` custom resource is provided through a `KafkaRebalance` custom resource defined as a "template".
+That is a `KafkaRebalance` custom resource with the `strimzi.io/rebalance: template` annotation set.
+When it is created, the `KafkaRebalanceAssemblyOperator` doesn't run any rebalancing.
+This is because it doesn't represent an "actual" rebalance request to get an optimization proposal, but it's just the place where configuration related to auto-rebalancing is defined.
+The user can specify rebalancing goals and options, such as `skipHardGoalCheck`, within the resource.
+When such a "template" `KafkaRebalance` custom resource is referenced within the `spec.cruiseControl.autoRebalance.config` field for one or more modes, the operator asks to Cruise Control for an optimization proposal with such configuration.
+If the `mode` and `brokers` fields are set, they will be just ignored because the resource doesn't represent an "actual" rebalancing.
```

Comment:
Why `rebalanceDisk` should be ignored? I am ignoring `mode` and `brokers` because these fields are implied by the scaling (up = add brokers mode, down = remove brokers mode) and the nodes being added and removed (during scaling). All other fields from a `KafkaRebalance` can be used instead.

Reply:
And how would you add or remove brokers with intra-broker balancing? I think you have to ignore it otherwise such autorebalance would be useless (in general, I guess it doesn't make sense to mix add / remove brokers with intra broker baůancing).

Reply:
Yeah you are right. What's the point of rebalancing within a broker which is going to be removed :-)